### ISSUE TITLE
Include subhub in dev environment

### ISSUE THREAD

**Issue by ianb** [mozilla/fxa#1082]
We could include [subhub](https://github.com/mozilla/subhub) in our dev setup.

It needs a non-public Stripe test API key to run, so the out-of-the-box experience for contributors wouldn't really work, but it's unrelated to what most contributors would be working on. It can run as its own server, but needs a DynamoDB(-like) database. So it would probably be two new Docker images.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/FXA-85)
┆Issue Number: FXA-85

**Comment by lmorchard** [comment]
Oh, huh, I thought subhub was in a private repo. Maybe we could gin up some deployment of it that has AWS stuff stubbed out and use a Stripe testing key for better local dev?

**Comment by lmorchard** [comment]
I started working through a manual install & setup of subhub, came up with some rough docs in a PR: https://github.com/mozilla/subhub/issues/43

One gotcha I can think of so far is that, for local dev, we'll need to somehow acquire Stripe API keys. For core team members, that's probably a Slack message to a colleague to get keys to the shared account. For any other contributor, they'll need to sign up for Stripe to get their own test keys.

Either way, I don't think we'll be able to include ready-to-use Stripe credentials in the repo. So, this will need a small manual step on setting up FxA local dev. Should probably be an optional step, leaving the default out-of-box state of local dev using in-memory stubs for subhub

**Comment by lmorchard** [comment]
Some progress on this...

TL;DR: You should be able to fire up a local subhub instance like so (swapping in your Stripe API key):
```
docker run -ti --rm -p 8012:8012 -e STRIPE_API_KEY=sk_test_TESTING_STRIPE_KEY lmorchard/subhub:dev
```
And change your local `packages/fxa-auth-server/config/dev.json` so that the `subhub` section looks like this:
```json
 "subhub": {
    "enabled": "true",
    "url": "http://127.0.0.1:8012/",
    "key": "abcde"
  },
```
You can leave the `stubs` section, as long as `useStubs` is false.

I pushed my Docker container to the hub, but that's probably not a great permanent place for it. This is the Dockerfile I came up with so far:
```Dockerfile
FROM python:3.7-alpine

RUN apk add --update gcc libc-dev linux-headers build-base python3-dev libffi-dev openssl-dev nodejs git \
    && pip install awscli

COPY ./requirements.txt requirements.txt
RUN pip install -r requirements.txt

RUN addgroup -g 10001 app && \
    adduser -D -G app -h /app -u 10001 app
WORKDIR /app

USER app

COPY . /app

RUN doit venv
RUN /app/subhub/.venv/bin/python -m setup develop

ENV SKIP_TESTS=yes \
    LOCAL_FLASK_PORT=8012 \
    DYNALITE_PORT=8011 \
    PAYMENT_API_KEY=abcde \
    SUPPORT_API_KEY=fghji

EXPOSE 8012
ENTRYPOINT ["doit", "local"]
```

**Comment by lmorchard** [comment]
Circling back to this and some things have changed. Pushed a new version of the container to Docker Hub and I made some updates to my Dockerfile:

```docker
FROM python:3.7-alpine

RUN apk add --update gcc libc-dev linux-headers build-base python3-dev libffi-dev openssl-dev nodejs git curl yarn \
    && pip install awscli

COPY ./requirements.txt requirements.txt
RUN pip install -r requirements.txt

RUN addgroup -g 10001 app && \
    adduser -D -G app -h /app -u 10001 app
WORKDIR /app

USER app

COPY . /app

RUN doit venv

RUN /app/venv/bin/python -m setup develop

# HACK: if building this container from a Windows host, some files end up
# chmod +x and that runs afoul of the server framework
USER root
RUN chmod -x /app/subhub/app.py
USER app

ENV SKIP_TESTS=yes \
    LOCAL_FLASK_PORT=8012 \
    DYNALITE_PORT=8011 \
    PAYMENT_API_KEY=abcde \
    SUPPORT_API_KEY=fghji

EXPOSE 8012
ENTRYPOINT ["doit", "local"]
```

**Comment by lmorchard** [comment]
This is a terrible way to do version control, but another tweak to Dockerfile after subhub changes renamed `requirements.txt` to `automation_requirements.txt`:
```dockerfile
FROM python:3.7-alpine

RUN apk add --update gcc libc-dev linux-headers build-base python3-dev libffi-dev openssl-dev nodejs git curl yarn \
    && pip install awscli

COPY ./automation_requirements.txt automation_requirements.txt
RUN pip install -r automation_requirements.txt

RUN addgroup -g 10001 app && \
    adduser -D -G app -h /app -u 10001 app
WORKDIR /app

USER app

COPY . /app

RUN doit venv

RUN /app/venv/bin/python -m setup develop

# HACK: if building this container from a Windows host, some files end up
# chmod +x and that runs afoul of the server framework
USER root
RUN chmod -x /app/subhub/app.py
USER app

ENV SKIP_TESTS=yes \
    LOCAL_FLASK_PORT=8012 \
    DYNALITE_PORT=8011 \
    PAYMENT_API_KEY=abcde \
    SUPPORT_API_KEY=fghji

EXPOSE 8012
ENTRYPOINT ["doit", "local"]               
```

**Comment by bowlofstew** [comment]
We have rolled out this on our side but wanted to get your feedback on it to see if it meets your needs. The docker file is available here: https://github.com/mozilla/subhub/blob/master/Dockerfile and the docker-compose to encapsulate calls, etc here: https://github.com/mozilla/subhub/blob/master/docker-compose.yml

**Comment by data-sync-user** [comment]
➤ Wil Clouser commented:

I'll close this due to architecture changes

**Comment by data-sync-user** [comment]
➤ Wil Clouser commented:

won't do
